T6275 implement callback/lava

gctucker opened the pull request:
This is to implement the /callback/lava API to create a boot entry into the database and save JSON meta-data and kernel log files using the data received from a LAVA v2 job.

khilman commented on app/utils/lava_log_parser.py:
```diff
@@ -0,0 +1,115 @@
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+# Based on log2html.py:
+# https://git.linaro.org/people/kevin.hilman/build-scripts.git/blob_plain/HEAD:/log2html.py
```
My git repos at Linaro are long dead.  The right URL for that mess, er, script is: 
https://github.com/khilman/lab/blob/master/log2html.py

Otherwise, really nice to see this going into the backend.  Thanks!

mattface replied:
@gctucker can you update this url, and @MiloCasagrande can you get this running on staging-api?

gctucker commented on app/utils/lava_log_parser.py:
```diff
@@ -0,0 +1,115 @@
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+# Based on log2html.py:
+# https://github.com/khilman/lab/blob/master/log2html.py
```
@khilman thanks, I've updated the URL

mattface commented:
You need to add pyyaml to the requirements.txt

MiloCasagrande commented on app/handlers/callback.py:
```diff
@@ -146,5 +148,7 @@ class LavaCallbackHandler(CallbackHandler):
-        pass
+    def _execute_callback(self, json_obj, lab_name):
+        taskq.lava_boot.apply_async(
+            [json_obj, lab_name, self.settings['dboptions']],
```
Passing `dboptions` here can be dropped.

MiloCasagrande commented on app/utils/callback/lava.py:
```diff
@@ -0,0 +1,234 @@
+
+LAVA_JOB_RESULT = {
+    COMPLETE: 'PASS',
+    INCOMPLETE: 'FAIL',
```
Nitpick: for consistency, please use double quotes, they have been used all around the code base.

MiloCasagrande commented on app/utils/callback/lava.py:
```diff
@@ -0,0 +1,234 @@
+    :type job_data: dict
+
+    """
+    meta['board_instance'] = job_data['actual_device_id']
```
Please use double quotes here as well, and in the code that follows.

MiloCasagrande commented on app/utils/callback/lava.py:
```diff
@@ -0,0 +1,234 @@
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+import bson
+import datetime
```
`bson` and `datetime` appear to be unused imports. If so, please remove them.

MiloCasagrande commented on app/utils/callback/lava.py:
```diff
@@ -0,0 +1,234 @@
+import datetime
+import errno
+import json
+import models.boot
```
Same here for `json` and `models.boot`: unused imports.

MiloCasagrande commented on app/utils/callback/lava.py:
```diff
@@ -0,0 +1,234 @@
+    INCOMPLETE: 'FAIL',
+}
+
+def _get_definition_meta(meta, job_data):
```
Need to add one line above this one (PEP8 fixes).

MiloCasagrande commented on app/utils/callback/lava.py:
```diff
@@ -0,0 +1,234 @@
+    job_meta = definition['metadata']
+    meta.update({x: job_meta[y] for x, y in meta_data_map.iteritems()})
+
+def _get_lava_boot_meta(meta, boot_meta):
```
Need to add one line above this one (PEP8 fixes).

MiloCasagrande commented on app/utils/callback/lava.py:
```diff
@@ -0,0 +1,234 @@
+    if kernel_messages:
+        meta['boot_warnings'] = kernel_messages
+
+def _get_lava_bootloader_meta(meta, bl_meta):
```
Need to add one line above this one (PEP8 fixes).

MiloCasagrande commented on app/utils/callback/lava.py:
```diff
@@ -0,0 +1,234 @@
+        elif step['name'] == 'bootloader-overlay':
+            _get_lava_bootloader_meta(meta, step['metadata'])
+
+def _add_boot_log(meta, log, base_path, errors):
```
Need to add one line above this one (PEP8 fixes).

MiloCasagrande commented on app/utils/callback/lava.py:
```diff
@@ -0,0 +1,234 @@
+
+    return ret_val
+
+def add_boot(job_data, lab_name, db_options, base_path=utils.BASE_PATH):
```
Need to add one line above this one (PEP8 fixes).

MiloCasagrande commented on app/utils/lava_log_parser.py:
```diff
@@ -0,0 +1,115 @@
+<h1>Boot log: {board}</h1>
+"""
+
+def run(log, boot, txt, html):
```
Need to add one line above this one (PEP8 fixes).

MiloCasagrande commented on app/utils/lava_log_parser.py:
```diff
@@ -0,0 +1,115 @@
+    html.write(log_buffer.getvalue())
+    html.write("</pre></font></body></html>\n")
+
+def main(args):
```
Need to add one line above this one (PEP8 fixes).

MiloCasagrande replied:
Do we still need to call this as a script from the command line?

gctucker replied:
Again, this was ported from log2html.py which can be called from the command line.  I originally thought we could have the same code to parse logs coming from v1 and v2 jobs, with the script used both here and in lava-ci.  It now looks like they are going to go down their own separate routes so I think we could drop the CLI support.

MiloCasagrande commented on app/utils/lava_log_parser.py:
```diff
@@ -0,0 +1,115 @@
+    with open(args.txt, 'w') as txt, open(args.html, 'w') as html:
+        run(log, boot, txt, html)
+
+if __name__ == '__main__':
```
Need to add one line above this one (PEP8 fixes).

MiloCasagrande commented on app/utils/callback/lava.py:
```diff
@@ -0,0 +1,234 @@
+        ret_code = 404
+        utils.LOG.error(msg)
+        utils.errors.add_error(errors, ret_code, msg)
+        return ret_code, doc_id, errors
```
No need to refactor the code here, but I prefer single exit point functions, even if the code will result in too much indentation. It won't become less readable, and it follows the code flow.

gctucker replied:
Yes, I think I originally wanted to throw an exception here but then changed my mind as it's really validating the data sent from the client (LAVA).  I'll address the other issues first and see if this can be tidied up easily.

gctucker replied:
I've pushed an updated version which deals better with catching exceptions, as well as having a single return path.

MiloCasagrande commented on app/utils/lava_log_parser.py:
```diff
@@ -0,0 +1,115 @@
+import os
+import re
+import sys
+import subprocess
```
`os`, `re`, `sys` and `subprocess` are unused imports.

MiloCasagrande commented on app/utils/lava_log_parser.py:
```diff
@@ -0,0 +1,115 @@
+            txt.write('\n')
+
+    html.write(HTML_HEAD.format(board=boot['board']))
+    html.write("<font size=-2><pre>\n")
```
Using the `<font>` tag should be avoided, it is an obsolete tag and it was deprecated. Browsers support for it can be dropped anytime.

Since it is only applied to the `<pre>` tag and there is a style block at the top, add something there: `pre { font-size: 80% }`

MiloCasagrande replied:
Stylistically speaking, why do we want everything here into a `<pre>` block, including the headers like `Boot result` etc...? I think only the real log text would make sense to be included in there.

gctucker replied:
I just ported @khilman 's log2html.py script to generate the same HTML output...  Can improve that with another patch.

MiloCasagrande commented on app/utils/callback/lava.py:
```diff
@@ -0,0 +1,234 @@
+    definition = yaml.load(job_data['definition'], Loader=yaml.CLoader)
+    meta['mach'] = definition['device_type']
+
+    meta_data_map = {
```
I think you can move this `meta_data_map` outside of the function definition, since it's always going to be the same and is not created on the fly.

Also, I would use as its keys the values from the `models` module to avoid typos or else:
```
models.DEFCONFIG_KEY: "kernel.defconfig_base",
models.DEFCONFIG_FULL_KEY: "kernel.defconfig",
...
```

MiloCasagrande commented on app/utils/callback/lava.py:
```diff
@@ -0,0 +1,234 @@
+import models.boot
+import os
+import utils
+import yaml
```
Need to add the `pyyaml` dependency in the `requirements.txt` file as well.

gctucker commented:
Thanks @MiloCasagrande for the review, and sorry I didn't run PEP8 on my changes... will do this.  I've also pushed a couple of fixes in the meantime.

gctucker commented:
@MiloCasagrande my branch is based on the master branch, I thought all your previous work on callback had already been merged in master?

MiloCasagrande commented:
No, there were a few changes added to the callback branch that were not
merged into master.
I started working on parsing the metadata from the LAVA payload, there was
no work on the logs.

-- 
Milo Casagrande
Linaro.org <www.linaro.org> │ Open source software for ARM SoCs

MiloCasagrande commented:
I don't think it's necessary to rebase your branch though, it's more
than OK like this and the work on the other branch was not ready
anyway.

-- 
Milo Casagrande
Linaro.org <www.linaro.org> │ Open source software for ARM SoCs

gctucker commented:
OK thanks, I've addressed all the comments apart from the HTML formatting and single return exit point.

I think it's worth testing this on staging to try and catch any functional errors and I'll clean up the rest while it's running.

MiloCasagrande commented:
I think it has already been deployed on staging, but I didn't do it though.

-- 
Milo Casagrande
Linaro.org <www.linaro.org> │ Open source software for ARM SoCs

mattface commented:
Yes I'm deploying it

gctucker commented:
Results:
https://staging.kernelci.org/boot/all/job/gtucker/branch/kernelci-test/kernel/kernelci-test-002-6-g1209cde722bc/

I believe there are still more failing jobs than there should be, some configs need to be investigated to determine whether it's really a kernel/board issue or something with how the v2 jobs were generated and submitted.

@MiloCasagrande  I've pushed an updated version of lava_log_parser.py, the output now looks like this:
https://people.collabora.com/~gtucker/tmp/kernelci-pass.html
https://people.collabora.com/~gtucker/tmp/kernelci-fail.html

mattface commented:
Yes as noted on IRC, I need to have a look at the device map because my 32 bit P4 board is getting booted with 64 bit kernels and will fail all of them. My x86-x5-z8350 (up board) is just failing everything so i've turned it off.

MiloCasagrande commented:
Logs look good.
Do you have the TXT format of those? In the <pre> section an HTML tag is
being created out of an email address. Wondering how is that being parsed...

-- 
Milo Casagrande
Linaro.org <www.linaro.org> │ Open source software for ARM SoCs

gctucker commented:
@MiloCasagrande yes here are the matching text versions:
https://people.collabora.com/~gtucker/tmp/kernelci-fail.txt
https://people.collabora.com/~gtucker/tmp/kernelci-pass.txt

gctucker commented:
I see your point, I think these special HTML characters need to be escaped in the <pre> section.

gctucker commented on app/utils/lava_log_parser.py:
```diff
@@ -0,0 +1,126 @@
+    for title, value in results.iteritems():
+        html.write("<li class=\"result\">{}: {}</li>".format(title, value))
+    html.write("</ul><pre>\n")
+    html.write(cgi.escape(log_buffer.getvalue()))
```
@MiloCasagrande Calling cgi.escape solves a lot of issues with special HTML characters:
https://people.collabora.com/~gtucker/tmp/kernelci-fail-escaped.html
Before:
https://people.collabora.com/~gtucker/tmp/kernelci-fail.html

MiloCasagrande commented:
On Wed, Jul 12, 2017 at 6:29 PM, Guillaume Tucker
<notifications@github.com> wrote:
>
> @MiloCasagrande Calling cgi.escape solves a lot of issues with special HTML characters:
> https://people.collabora.com/~gtucker/tmp/kernelci-fail-escaped.html
> Before:
> https://people.collabora.com/~gtucker/tmp/kernelci-fail.html

That's great, but you lose some of the formatting now.
Look at all the escaped "span" tags at the top and at the bottom.

Not sure if it will work, but you probably need to apply the escape
function when you create each line of the log_buffer (the StringIO
object), calling it only on the content, not on the formatting tags.

-- 
Milo Casagrande
Linaro.org <www.linaro.org> │ Open source software for ARM SoCs

gctucker commented:
Oops, yes sorry didn't notice that.  Escaping each line does work:
https://people.collabora.com/~gtucker/tmp/kernelci-fail-escaped-2.html
The text version is the same as before:
https://people.collabora.com/~gtucker/tmp/kernelci-fail-2.txt

PS Also replaced some quotes with double quotes in lava_log_parser.py for consistency.
